File: 5.2 Ad Hoc Mathematics/Simulation Harder/p616.py
# Competitive Programming 3
# Problem 616

while True:
    N = int(input())
    if N < 0: break

    print('{} coconuts, '.format(N), end='')
    works = False
    for p in range(int(N**0.5 + 0.5), 1, -1):
        works = True
        n = N
        
        for _ in range(p):
            take = n//p
            rem = n % p
            if rem != 1:
                works = False
                break
            else:
                n -= take + 1
            
        if n % p != 0:
            works = False
            
        if works: break
    
        
    if not works: print('no solution')
    else: print('{} people and 1 monkey'.format(p))

File: 5.2 Ad Hoc Mathematics/Simulation Harder/p616.cpp
// Competitive Programming 3
// Problem 616

#include <iostream>
#include <cmath>
using namespace std;

int main(){
    while (1){
        int N;
        cin >> N;
        if (N < 0) break;

        bool works = false;
        int p;

        for (p = (int)(sqrt(N)+0.5); p > 1; p--){
            works = true;
            int n = N;

            for (int q = 0; q < p; q++){
                int take = n / p;
                int rem = n % p;

                if (rem != 1){
                    works = false;
                    break;
                }
                else
                    n -= take + 1;
            }

            if (n % p != 0)
                works = false;

            if (works) break;
        }

        printf("%d coconuts, ", N);
        if (!works) printf("no solution\n");
        else printf("%d people and 1 monkey\n", p);
    }
    return 0;
}
